### ISSUE TITLE
[MemberNotNull] isn't respected on local functions

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn#56256]
**Version Used**: 
9525fa5

**Steps to Reproduce**:
https://sharplab.io/#v2:EYLgtghglgdgPgYhgVwDaosVBTABNmTHAWACgABABl3IEYA6AESggHMYB7AZwBcoBjLvQDCHACbYAgoVQBPLlC4BuMmXIBmGgCZcwsgG8yuY7gAOAJygA3CDzwdgAK2z8eAflwB9B45WkTuEYmGriwPLgAsgAUAJRBxob+ASYAkjBQPLF+ycEA7F4+9ADi2DwAEhBcABaiElnxyQ0BANoR2GDA2OYAchw83WioUYRg2BwAZlHeTjExALpNwQAsuGkZsbgAvAB8BU5buDDYAO64Pi6ZMdkmAL5kN0A===
```C#
#nullable enable
using System.Diagnostics.CodeAnalysis;

public class C
{
    private object? _obj;
    
    public int M()
    {
        Init();
        return _obj.GetHashCode();
        
        [MemberNotNull(nameof(_obj))]
        void Init() => _obj = new object();
    }
}
````

**Expected Behavior**:
No warning.

**Actual Behavior**:
`warning CS8602: Dereference of a possibly null reference.`

**Comment by stephentoub** [comment]
Ah, dup of https://github.com/dotnet/roslyn/issues/47896

**Comment by stephentoub** [comment]
Apparently I opened the same issue last year ;-)
https://github.com/dotnet/roslyn/issues/49234

**Comment by Youssef1313** [comment]
@stephentoub #47896 was closed as fixed, but this issue still repros. This issue wasn't probably an exact dupe of #47896 and should be reopened. Can you reopen? Thanks!

**Comment by RikkiGibson** [comment]
LDM did decide that we should do this and that `MemberNotNull` applied to the function applies to members of the captured `this`.

**Comment by jcouv** [comment]
Implemented in 17.13 :-)
